hnswlib/hnswalg.h

```diff
@@ -974,11 +1146,99 @@ namespace hnswlib
             elements_to_persist_.clear();
         }
 
-        void loadPersistedIndex(SpaceInterface<dist_t> *s, size_t max_elements_i = 0)
+        constexpr size_t calculateHeaderSize() const {
+            #define SIZE_ACTION(field) + sizeof(field)
+            return 0 HEADER_FIELDS(SIZE_ACTION);
+            #undef SIZE_ACTION
+        }
+
+        size_t calculateLinkListSize() {
+            size_t total_size = 0;
+            for (size_t i = 0; i < cur_element_count; i++) {
+                unsigned int linkListSize = element_levels_[i] > 0 ? size_links_per_element_ * element_levels_[i] : 0;
+                total_size += sizeof(unsigned int) + linkListSize;
+            }
+            return total_size;
+        }
+
+        size_t calculateDataLevel0Size() {
+            return max_elements_ * size_data_per_element_;
+        }
+
+        size_t calculateLengthSize() {
+            return max_elements_ * sizeof(float);
+        }
+
+        void serializeHeaderToBuffer(char* buffer, size_t buffer_size) {
+            char *init_buffer = buffer;
+            #define WRITE_ACTION(field) \
+                do { memcpy(buffer, &(field), sizeof(field)); buffer += sizeof(field); } while (0);
+            HEADER_FIELDS(WRITE_ACTION)
+            #undef WRITE_ACTION
+        }
+
+        void serializeDataLevel0ToBuffer(char* buffer, size_t buffer_size) {
+            memcpy(buffer, data_level0_memory_, max_elements_ * size_data_per_element_);
+        }
+
+        void serializeLengthToBuffer(char* buffer, size_t buffer_size) {
+            memcpy(buffer, length_memory_, max_elements_ * sizeof(float));
+        }
+
+        void serializeLinkListsToBuffer(char* buffer, size_t buffer_size) {
+            for (size_t i = 0; i < cur_element_count; i++) {
+                unsigned int linkListSize = element_levels_[i] > 0 ? size_links_per_element_ * element_levels_[i] : 0;
+                memcpy(buffer, &linkListSize, sizeof(unsigned int));
+                buffer += sizeof(unsigned int);
+                if (linkListSize) {
+                    memcpy(buffer, linkLists_[i], linkListSize);
+                    buffer += linkListSize;
+                }
+            }
+        }
+
+        // Memory-based persistence functions
+        HnswData* serializeToHnswData() {
```

Comment:
Would it make more sense to have the caller supply buffers this can write into, this way all memory allocation and tracking is done in rust